lib/core/src/Cardano/Wallet/DB/Sqlite/AddressBook.hs

```diff
@@ -99,56 +81,47 @@ instance AddressBookIso (Seq.SeqState n k)
         in  iso from2 to2
 
 -- | Isomorphism for sequential address book.
-instance 
-    ( MkKeyFingerprint key (Proxy n, key 'AddressK XPub)
-    , GetPurpose key
-    , SoftDerivation key
-    , PaymentAddress n key
-    , (key == SharedKey) ~ 'False
-    ) => AddressBookIso (Seq.SeqState n key)
+instance ( (key == SharedKey) ~ 'False ) => AddressBookIso (Seq.SeqState n key)
   where
     data Prologue (Seq.SeqState n key)
         = SeqPrologue (Seq.SeqState n key)
         -- Trick: We keep the type, but we empty the discovered addresses
     data Discoveries (Seq.SeqState n key)
         = SeqDiscoveries
-            (SeqAddressList 'UtxoInternal)
-            (SeqAddressList 'UtxoExternal)
+            (SeqAddressMap 'UtxoInternal key)
+            (SeqAddressMap 'UtxoExternal key)
 
     addressIso = iso from to
       where
-        from (Seq.SeqState int ext a b c) =
-            ( SeqPrologue $ Seq.SeqState (emptyPool int) (emptyPool ext) a b c
-            , SeqDiscoveries (toDiscoveries @n int) (toDiscoveries @n ext)
-            )
-        to (SeqPrologue (Seq.SeqState int ext a b c), SeqDiscoveries ints exts)
-          = Seq.SeqState (fromDiscoveries @n int ints) (fromDiscoveries @n ext exts) a b c
-
--- | Extract the discovered addresses from an address pool.
-toDiscoveries
-    :: forall (n :: NetworkDiscriminant) (c :: Role) key.
-    ( GetPurpose key
-    , PaymentAddress n key
-    , Typeable c
-    ) => Seq.AddressPool c key -> SeqAddressList c
-toDiscoveries pool = SeqAddressList
-    [ (a,st) | (a,st,_) <- Seq.addresses (liftPaymentAddress @n) pool ]
-
--- | Fill an empty address pool with addresses.
-fromDiscoveries
-    :: forall (n :: NetworkDiscriminant) (c :: Role) key.
-    ( MkKeyFingerprint key (Proxy n, key 'AddressK XPub)
-    , MkKeyFingerprint key Address
-    , SoftDerivation key
-    , Typeable c
-    ) => Seq.AddressPool c key -> SeqAddressList c -> Seq.AddressPool c key
-fromDiscoveries ctx (SeqAddressList addrs) =
-    Seq.mkAddressPool @n (Seq.context ctx) (Seq.gap ctx) addrs
-
--- | Remove all discovered addresses from an address pool,
--- but keep context.
-emptyPool :: Seq.AddressPool c key -> Seq.AddressPool c key
-emptyPool pool = pool{ Seq.indexedKeys = Map.empty }
+        from (Seq.SeqState int ext a b c d) =
+            let int0 = clear int
```

Comment:
what about introducing "AddressPool.emptyPool" instead of clearing?

Reply:
Well, `AddressPool.new` constructs an empty pool, but it needs a `generator` and a `gap` as arguments. The function `AddressPool.clear` essentially just calls `AddressPool.new` with these arguments taken from an existing pool. Unfolding the definition of `clear` here would yield

```haskell
let int0 = AddressPool.new (AddressPool.generator int) (AddressPool.gap int) 
```

I also find it unfortunate that the pool parameters (`gap`, `generator`), which belong to the `Prologue` part, are baked in the abstract data type together with the pool addresses, which belong to the `Discoveries` part, but I think that it's a necessary compromise to keep the invariants intact. At some point in the code, things will look a bit weird when trying to "unbake" the two, and, well, this is that point in the code.

Reply:
ok. thanks for the clarification. sounds reasonable